src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -571,97 +599,189 @@ private boolean createFeedbackForDuplicateTests(Result result, ProgrammingExerci
+        }).sum();
+
+        /*
+         * The points are capped by the maximum achievable points. The cap is applied before the static code analysis penalty is subtracted as otherwise the penalty won't have any
+         * effect in some cases. For example with maxPoints=20, successfulTestPoints=30 and penalty=10, a student would still receive the full 20 points, if the points are not
+         * capped before the penalty is subtracted. With the implemented order in place successfulTestPoints will be capped to 20 points first, then the penalty is subtracted
+         * resulting in 10 points.
+         */
+        double maxPoints = programmingExercise.getMaxPoints() + Optional.ofNullable(programmingExercise.getBonusPoints()).orElse(0.0);
+
+        if (Double.isNaN(successfulTestPoints)) {
+            successfulTestPoints = 0;
+        }
+        else if (successfulTestPoints > maxPoints) {
+            successfulTestPoints = maxPoints;
         }
 
-        result.getFeedbacks().forEach(feedback -> {
-            if (feedback.getCredits() == null) {
-                feedback.setCredits(0D);
-            }
-        });
+        return successfulTestPoints;
+    }
+
+    /**
+     * Updates the feedback corresponding to the test case with the given credits.
+     * @param result which should be updated.
+     * @param testCase the feedback that should be updated corresponds to.
+     * @param credits that should be set in the feedback.
+     */
+    private void setCreditsForTestCaseFeedback(final Result result, final ProgrammingExerciseTestCase testCase, double credits) {
+        // We need to compare testcases ignoring the case, because the testcaseRepository is case-insensitive
+        result.getFeedbacks().stream().filter(fb -> FeedbackType.AUTOMATIC.equals(fb.getType()) && fb.getText().equalsIgnoreCase(testCase.getTestName())).findFirst()
+                .ifPresent(feedback -> feedback.setCredits(credits));
+    }
+
+    /**
+     * Calculates the points that should be awarded for a successful test case.
+     * @param result used to determine if the calculation is performed for the solution.
+     * @param programmingExercise the result belongs to.
+     * @param test for which the points should be calculated.
+     * @param totalTestCaseCount in the given exercise.
+     * @param weightSum of all test cases in the exercise.
+     * @return the points which should be awarded for successfully completing the test case.
+     */
+    private double calculatePointsForSuccessfulTestCase(final Result result, final ProgrammingExercise programmingExercise, final ProgrammingExerciseTestCase test,
+            int totalTestCaseCount, double weightSum) {
+        final boolean isWeightSumZero = Precision.equals(weightSum, 0, 1E-8);
+        final double testPoints;
+
+        // a non-zero score has to be calculated for the solution even if the weight sum is zero
+        // to avoid a warning to the instructor that the solution has faults
```

Comment:
Not quite sure what you mean by that, shouldn't it be good that an instructor knows if a solution has faults?
Please specify

Reply:
The comment was indeed to short. I extended the explanation now to make clear in which cases the warning should be shown and how this only avoids false warning.